Add AXI4 interface

sequencer opened the pull request.

sequencer committed:
add axi4 interface definition

cyyself commented on axi4/src/axi4/bundle/ChiselType.scala:
```diff
@@ -0,0 +1,124 @@
+  val addr: UInt = UInt(addrWidth.W)
+  val len: UInt = UInt(8.W)
+  val size: UInt = UInt(3.W)
+  val burst: UInt = UInt(2.W)
```
For `burst` and `lock`, it is better to use an enum-like definition.

sequencer replied:
yeah, enum... enum... Actually `ChiselEnum` is the least thing I wanna use, but I'll have a try.

sequencer replied:
It's almost impossible to describe it in Chisel. I added the helper 7348efe2506dbdfccafe04d6692857fc4276812f

cyyself commented on axi4/src/axi4/bundle/ChiselType.scala:
```diff
@@ -0,0 +1,124 @@
+  val len: UInt = UInt(8.W)
+  val size: UInt = UInt(3.W)
+  val burst: UInt = UInt(2.W)
+  val lock: Bool = Bool()
```
For `lock`, `cache`, `prot`, `qos`, we should add Option here since it does not appear in most of the IPs. A solution might be adding some parameters like `hasLock`, `hasCache`, etc.

sequencer replied:
Good catch. I didn't notice these optional signals. See specification:
> Optional Outputs
If a component might require a value that does not match the default value, then the component must have the output signal present.
If a component always requires the value that matches the default value, specified in Default signal values on page A9-110, then it is not required that the component has the signal present.
Optional Inputs
An input signal can be omitted if the Manager or Subordinate does not need to observe the input signal for correct functional operation.

Basically there are two questions here:
1. how to connect he default signal:
There are some references:
see RocketChip has a feature via [`BundleMap`](https://github.com/chipsalliance/rocket-chip/blob/dbcb06afe1c76d1129cb6d264949322a34c37185/src/main/scala/util/BundleMap.scala#L123)
Chisel provide a feature via custom `Connectable` chipsalliance/chisel#3978
2. if AXI Bundles are in different parameters. Do we allow to connect them together? 
 - if yes, there might be some implicit truncation and expansion or connect based on the default value. This is some how unsafe from my mind, but will be super convenient for interconnection design.
 - if no, then we don't have such issues, but how to design the explicit API to convert and connect different AXI types?

sequencer replied:
The specification says:
> This specification suggests that, in general, for maximum IP reuse, an AXI component interface includes all signals. The presence of all signals reduces the risk of error at the system integration phase of the design flow and it can also help support some design flows that do not effectively support default values for absent signals.

I prefer Chisel goes to the full signal way based on this suggestion.
However we are exposing `hasRegion`, `hasId`, `hasLen` and other configurations, but keep the interface unchanged. during the connection, these optional will be ignored and set to default value. And ask EDA tool to do automatically optimization.

sequencer replied:
Issue filed at chipsalliance/chisel#4068, link this PR to it.

sequencer committed:
add optional field parameter for future use

These are the type of Interface, defining the connect behavior, which
will be used for customConnectable in the future PR. If some support is
set to false, the connection will be ignored and connect to default type
based on the specification.

sequencer committed:
add enum helper for burst, lock, cache and prot.
